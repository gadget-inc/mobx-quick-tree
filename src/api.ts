import type {
  IDisposer,
  IAnyComplexType as MSTAnyComplexType,
  IAnyModelType as MSTAnyModelType,
  IAnyType as MSTAnyType,
  IStateTreeNode as MSTStateTreeNode,
} from "mobx-state-tree";
import {
  applySnapshot as mstApplySnapshot,
  flow as mstFlow,
  getEnv as mstGetEnv,
  getParent as mstGetParent,
  getParentOfType as mstGetParentOfType,
  getRoot as mstGetRoot,
  getType as mstGetType,
  hasEnv as mstHasEnv,
  isArrayType as mstIsArrayType,
  isIdentifierType as mstIsIdentifierType,
  isMapType as mstIsMapType,
  isModelType as mstIsModelType,
  isReferenceType as mstIsReferenceType,
  isRoot as mstIsRoot,
  isStateTreeNode as mstIsStateTreeNode,
  isType as mstIsType,
  onSnapshot as mstOnSnapshot,
  resolveIdentifier as mstResolveIdentifier,
} from "mobx-state-tree";
import type { FlowReturn } from "mobx-state-tree/dist/internal";
import { CantRunActionError } from "./errors";
import { $context, $parent, $quickType, $readOnly, $type } from "./symbols";
import type {
  CreateTypes,
  IAnyComplexType,
  IAnyModelType,
  IAnyStateTreeNode,
  IAnyType,
  IArrayType,
  IClassModelType,
  IMapType,
  IReferenceType,
  IStateTreeNode,
  IType,
  Instance,
  SnapshotIn,
  TreeContext,
} from "./types";

export {
  addDisposer,
  addMiddleware,
  applyPatch,
  clone,
  createActionTrackingMiddleware2,
  destroy,
  detach,
  escapeJsonPath,
  getIdentifier,
  getPath,
  getPathParts,
  getRunningActionContext,
  hasParent,
  isActionContextThisOrChildOf,
  isAlive,
  isValidReference,
  joinJsonPath,
  onAction,
  onPatch,
  recordPatches,
  resolvePath,
  setLivelinessChecking,
  splitJsonPath,
  tryReference,
  typecheck,
  unescapeJsonPath,
  walk,
} from "mobx-state-tree";
export { ClassModel, action, extend, register, view, snapshottedView, volatile, volatileAction } from "./class-model";
export { getSnapshot } from "./snapshot";

export const isType = (value: any): value is IAnyType => {
  return $quickType in value;
};

/**
 * Returns true if the given object is a complex type in observable mode
 * @param value any object
 * @returns
 */
export const isStateTreeNode = (value: any): value is IStateTreeNode => {
  if (mstIsStateTreeNode(value)) {
    return true;
  }

  return typeof value === "object" && value !== null && $type in value;
};

/**
 * Returns true if the given object is a complex type in readonly mode
 * @param value any object
 * @returns
 */
export const isReadOnlyNode = (value: any): value is IStateTreeNode => {
  return typeof value === "object" && value !== null && $readOnly in value;
};

export const getParent = <T extends IAnyType>(value: IAnyStateTreeNode, depth = 1): Instance<T> => {
  if (mstIsStateTreeNode(value)) {
    return mstGetParent(value, depth) as Instance<T>;
  }

  while (value && depth > 0) {
    value = value[$parent];
    depth -= 1;
  }

  if (!value) {
    throw new Error("failed to get parent");
  }

  return value;
};

export function getParentOfType<T extends IAnyComplexType | MSTAnyComplexType>(value: IAnyStateTreeNode, type: T): Instance<T> {
  if (mstIsStateTreeNode(value)) {
    if (mstIsType(type)) {
      return mstGetParentOfType(value, type);
    } else {
      return mstGetParentOfType(value, type.mstType) as Instance<T>;
    }
  }

  value = value[$parent];
  while (value) {
    if (type.is(value)) {
      break;
    }

    value = value[$parent];
  }

  if (!value) {
    throw new Error("failed to get parent");
  }

  return value;
}

export function getType(value: IAnyStateTreeNode): MSTAnyComplexType | IAnyComplexType {
  if (mstIsStateTreeNode(value)) {
    return mstGetType(value);
  }

  return value[$type];
}

/** @hidden */
export function getContext(value: IAnyStateTreeNode): TreeContext | null {
  if (mstIsStateTreeNode(value)) {
    throw new Error("can't get the context of an observable node, this function is only for use on readonly nodes");
  }

  // Assumes no cycles, otherwise this is an infinite loop
  let currentNode: IStateTreeNode = value;
  while (currentNode) {
    const context = (currentNode as any)[$context];
    if (context !== undefined) {
      return context;
    }

    currentNode = (currentNode as any)[$parent];
  }

  return null;
}

export function getEnv<Env = any>(value: IAnyStateTreeNode): Env {
  if (mstIsStateTreeNode(value)) {
    return mstGetEnv(value);
  }

  const env = getContext(value)?.env;
  if (!env) {
    throw new Error(`Failed to find the environment of ${value}`);
  }

  return env as Env;
}

export function hasEnv(value: IAnyStateTreeNode): boolean {
  if (mstIsStateTreeNode(value)) {
    return mstHasEnv(value);
  }

  return !!getContext(value)?.env;
}

export const getRoot = <T extends IAnyType>(value: IAnyStateTreeNode): Instance<T> => {
  if (mstIsStateTreeNode(value)) {
    return mstGetRoot(value) as Instance<T>;
  }

  // Assumes no cycles, otherwise this is an infinite loop
  // eslint-disable-next-line no-constant-condition
  while (true) {
    const newValue = value[$parent];
    if (newValue) {
      value = newValue;
    } else {
      return value;
    }
  }
};

export const isRoot = (value: IAnyStateTreeNode): boolean => {
  if (mstIsStateTreeNode(value)) {
    return mstIsRoot(value);
  }

  return !value[$parent];
};

export function resolveIdentifier<T extends IAnyModelType>(
  type: T,
  target: IStateTreeNode<IAnyType>,
  identifier: string,
): Instance<T> | undefined {
  if (mstIsStateTreeNode(target)) {
    if (isType(type)) {
      return mstResolveIdentifier(type.mstType as MSTAnyModelType, target, identifier);
    } else {
      return mstResolveIdentifier(type, target, identifier);
    }
  }

  const context = getContext(target);
  if (!context) {
    throw new Error("can't resolve references in a readonly tree with no context");
  }
  return context.referenceCache.get(identifier) as Instance<T> | undefined;
}

export const applySnapshot = <T extends IAnyType>(target: IStateTreeNode<T>, snapshot: SnapshotIn<T>): void => {
  if (mstIsStateTreeNode(target)) {
    mstApplySnapshot(target as MSTStateTreeNode, snapshot);
    return;
  }

  throw new Error("can't apply a snapshot to a mobx-quick-tree node");
};

export const onSnapshot = <S>(
  target: IStateTreeNode<IType<any, S, any>> | IStateTreeNode<IClassModelType<any, any, S>>,
  callback: (snapshot: S) => void,
): IDisposer => {
  if (mstIsStateTreeNode(target)) {
    return mstOnSnapshot<S>(target as MSTStateTreeNode, callback);
  }

  throw new Error("can't use onSnapshot with a mobx-quick-tree node");
};

export const isArrayType = (value: IAnyType | MSTAnyType): value is IArrayType<IAnyType> => {
  if (mstIsType(value)) {
    return mstIsArrayType(value);
  }
  return mstIsArrayType(value.mstType);
};

export const isMapType = (value: IAnyType | MSTAnyType): value is IMapType<IAnyType> => {
  if (mstIsType(value)) {
    return mstIsMapType(value);
  }
  return mstIsMapType(value.mstType);
};

export const isModelType = (value: IAnyType | MSTAnyType): value is IAnyModelType => {
  if (mstIsType(value)) {
    return mstIsModelType(value);
  }
  return mstIsModelType(value.mstType);
};

export const isReferenceType = (value: IAnyType | MSTAnyType): value is IReferenceType<IAnyComplexType> => {
  if (mstIsType(value)) {
    return mstIsReferenceType(value);
  }
  return mstIsReferenceType(value.mstType);
};

export const isIdentifierType = (value: IAnyType | MSTAnyType): value is IReferenceType<IAnyComplexType> => {
  if (mstIsType(value)) {
    return mstIsIdentifierType(value);
  }
  return mstIsIdentifierType(value.mstType);
};

export function cast<O extends string | number | boolean | null | undefined = never>(snapshotOrInstance: O): O;
export function cast<O extends IAnyType>(snapshotOrInstance: CreateTypes<O>): O;
export function cast(snapshotOrInstance: never): never;
export function cast(snapshotOrInstance: any): any {
  return snapshotOrInstance;
}

/**
 * Defines a new asynchronous action. `mobx-quick-tree` (and `mobx-state-tree`) require this wrapper around asynchronous actions, and require those action functions to be generators using `yield` instead of `await`.
 *
 * Accepts an incoming generator function and returns a new async function with the right mobx-state-tree wrapping.
 * See https://mobx-state-tree.js.org/concepts/async-actions for more info.
 */
export function flow<R, Args extends any[], This = unknown>(
  generator: (this: This, ...args: Args) => Generator<PromiseLike<any>, R, any>,
): (...args: Args) => Promise<FlowReturn<R>> {
  // wrap the passed generator in a function which restores the correct value of `this`
  const wrappedGenerator = mstFlow(function* (args: Args, instance: This) {
    return yield* generator.call(instance, ...args);
  });

  // return an async function to set on the prototype which ensures the instance is not readonly
  return async function (this: This, ...args: Args): Promise<FlowReturn<R>> {
    if ((this?.constructor as any)?.isMQTClassModel) {
      throw new CantRunActionError(`Can't run flow action for a readonly instance`);
    }
    return await wrappedGenerator(args, this);
  };
}
