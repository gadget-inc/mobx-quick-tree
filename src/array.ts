import { isStateTreeNode, types } from "mobx-state-tree";
import { BaseType } from "./base";
import { ensureRegistered } from "./class-model";
import { getSnapshot } from "./snapshot";
import { $context, $parent, $readOnly, $type } from "./symbols";
import type { IAnyStateTreeNode, IAnyType, IArrayType, IMSTArray, IStateTreeNode, Instance, TreeContext } from "./types";
import { SafeReferenceType } from "./reference";

export class QuickArray<T extends IAnyType> extends Array<Instance<T>> implements IMSTArray<T> {
  static get [Symbol.species]() {
    return Array;
  }

  /** @hidden */
  readonly [$context]: any;
  /** @hidden */
  readonly [$parent]: IStateTreeNode | null;
  /** @hidden */
  readonly [$type]: [this] | [any];

  constructor(type: any, parent: IStateTreeNode | null, context: TreeContext, items?: Instance<T>[]) {
    super();
    this[$type] = type;
    this[$parent] = parent;
    this[$context] = context;
    if (items) {
      this.push(...items);
    }
  }

  get [Symbol.toStringTag]() {
    return "Array" as const;
  }

  get [$readOnly]() {
    return true;
  }

  spliceWithArray(_index: number, _deleteCount?: number, _newItems?: Instance<T>[]): Instance<T>[] {
    throw new Error("cannot spliceWithArray on a QuickArray instance");
  }

  clear(): Instance<T>[] {
    throw new Error("cannot clear a QuickArray instance");
  }

  replace(_newItems: Instance<T>[]): Instance<T>[] {
    throw new Error("cannot replace a QuickArray instance");
  }

  remove(_value: Instance<T>): boolean {
    throw new Error("cannot remove from a QuickArray instance");
  }

  toJSON(): Instance<T>[] {
    return getSnapshot(this);
  }
}

export class ArrayType<T extends IAnyType> extends BaseType<Array<T["InputType"]> | undefined, T["OutputType"][], IMSTArray<T>> {
  constructor(readonly childrenType: T) {
    super(types.array(childrenType.mstType));
  }

  is(value: IAnyStateTreeNode): value is this["InstanceType"];
  is(value: any): value is this["InputType"] | this["InstanceType"] {
    if (isStateTreeNode(value)) {
      return this.mstType.is(value);
    }

    if (value === undefined) {
      return true;
    }

    if (!Array.isArray(value) && !(value instanceof QuickArray)) {
      return false;
    }

    if ((value as any)[$type] === this) {
      return true;
    }

    return value.every((child: any) => this.childrenType.is(child));
  }

  instantiate(snapshot: this["InputType"] | undefined, context: TreeContext, parent: IStateTreeNode | null): this["InstanceType"] {
    const array = new QuickArray<T>(this, parent, context);
    if (snapshot) {
      let instances = snapshot.map((element) => this.childrenType.instantiate(element, context, array));
      if (this.childrenType instanceof SafeReferenceType && this.childrenType.options?.acceptsUndefined === false) {
        instances = instances.filter((instance) => instance != null);
      }
      array.push(...instances);
    }

    return array as this["InstanceType"];
  }

  async schemaHash() {
    return `array:${await this.childrenType.schemaHash()}`;
  }
}

export const array = <T extends IAnyType>(childrenType: T): IArrayType<T> => {
  ensureRegistered(childrenType);
  return new ArrayType(childrenType);
};
