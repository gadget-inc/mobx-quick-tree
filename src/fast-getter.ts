import type { PropertyMetadata, ViewMetadata } from "./class-model";
import { getPropertyDescriptor } from "./class-model";
import { RegistrationError } from "./errors";
import { $notYetMemoized, $readOnly } from "./symbols";

/** Assemble a function for getting the value of a readonly instance very quickly with static dispatch to properties */
export class FastGetBuilder {
  memoizableProperties: string[];

  constructor(
    metadatas: PropertyMetadata[],
    readonly klass: { new (...args: any[]): any },
  ) {
    this.memoizableProperties = metadatas
      .filter((metadata): metadata is ViewMetadata => {
        if (metadata.type !== "view" && metadata.type !== "snapshotted-view") return false;
        const property = metadata.property;
        const descriptor = getPropertyDescriptor(klass.prototype, property);
        if (!descriptor) {
          throw new RegistrationError(`Property ${property} not found on ${klass} prototype, can't register view for class model`);
        }
        return descriptor.get !== undefined;
      })
      .map((metadata) => metadata.property);
  }

  memoSymbolName(property: string) {
    return `mqt/${property}-memo`;
  }

  outerClosureStatements(className: string) {
    return this.memoizableProperties
      .map(
        (property) => `
          const ${property}Memo = Symbol.for("${this.memoSymbolName(property)}");
          ${className}.prototype[${property}Memo] = $notYetMemoized;
        `,
      )
      .join("\n");
  }

  buildGetter(property: string, descriptor: PropertyDescriptor) {
    const $memo = Symbol.for(this.memoSymbolName(property));
    const source = `
      (
        function build({ $readOnly, $memo, $notYetMemoized, getValue }) {
          return function get${property}(model, imports) {
            if (!this[$readOnly]) return getValue.call(this);
            let value = this[$memo];
            if (value !== $notYetMemoized) {
              return value;
            }

            value = getValue.call(this);
            this[$memo] = value;
            return value;
          }
        }
      )
      //# sourceURL=mqt-eval/dynamic/${this.klass.name}-${property}-get.js
    `;

    try {
      const builder = eval(source);
      return builder({ $readOnly, $memo, $notYetMemoized, getValue: descriptor.get });
    } catch (error) {
      console.error(`Error building getter for ${this.klass.name}#${property}`);
      console.error(`Compiled source:\n${source}`);
      throw error;
    }
  }
}
