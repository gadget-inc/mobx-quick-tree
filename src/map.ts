import type { IInterceptor, IMapDidChange, IMapWillChange, Lambda } from "mobx";
import { isStateTreeNode, types } from "mobx-state-tree";
import { BaseType } from "./base";
import { ensureRegistered } from "./class-model";
import { getSnapshot } from "./snapshot";
import { $context, $parent, $readOnly, $type } from "./symbols";
import type {
  CreateTypes,
  IAnyStateTreeNode,
  IAnyType,
  IMSTMap,
  IMapType,
  IStateTreeNode,
  Instance,
  TreeContext,
  SnapshotOut,
} from "./types";
import { SafeReferenceType } from "./reference";

export class QuickMap<T extends IAnyType> extends Map<string, Instance<T>> implements IMSTMap<T> {
  static get [Symbol.species]() {
    return Map;
  }

  /** @hidden */
  readonly [$context]?: any;
  /** @hidden */
  readonly [$parent]?: IStateTreeNode | null;
  /** @hidden */
  readonly [$type]?: [this] | [any];

  constructor(type: any, parent: IStateTreeNode | null, context: TreeContext) {
    super();
    this[$type] = type;
    this[$parent] = parent;
    this[$context] = context;
  }

  get [Symbol.toStringTag]() {
    return "Map" as const;
  }

  get [$readOnly]() {
    return true;
  }

  forEach(callbackfn: (value: Instance<T>, key: string, map: this) => void, thisArg?: any): void {
    super.forEach((value, key) => callbackfn(value, key, thisArg ?? this));
  }

  put(_value: CreateTypes<T>): Instance<T> {
    throw new Error("put not supported in QuickMap");
  }

  merge(_other: any): this {
    throw new Error("merge not supported in QuickMap");
  }

  replace(_values: any): this {
    throw new Error("replace not supported in QuickMap");
  }

  toJSON(): Record<string, SnapshotOut<T>> {
    return getSnapshot(this);
  }

  observe(_listener: (changes: IMapDidChange<string, Instance<T>>) => void, _fireImmediately?: boolean): Lambda {
    throw new Error("observer not supported in QuickMap");
  }

  intercept(_handler: IInterceptor<IMapWillChange<string, Instance<T>>>): Lambda {
    throw new Error("intercept not supported in QuickMap");
  }
}

export class MapType<T extends IAnyType> extends BaseType<
  Record<string, T["InputType"]> | undefined,
  Record<string, T["OutputType"]>,
  IMSTMap<T>
> {
  constructor(readonly childrenType: T) {
    super(types.map(childrenType.mstType));
  }

  is(value: IAnyStateTreeNode): value is this["InstanceType"];
  is(value: any): value is this["InputType"] | this["InstanceType"] {
    if (isStateTreeNode(value)) {
      return this.mstType.is(value);
    }

    if (value === undefined) {
      return true;
    }

    if (typeof value !== "object" || value === null) {
      return false;
    }

    if (!(value instanceof QuickMap) && Object.getPrototypeOf(value) != Object.prototype) {
      return false;
    }

    if (value[$type] === this) {
      return true;
    }

    const children = Object.values(value as object);
    return children.every((child) => this.childrenType.is(child));
  }

  instantiate(snapshot: this["InputType"] | undefined, context: TreeContext, parent: IStateTreeNode | null): this["InstanceType"] {
    const map = new QuickMap<T>(this, parent, context);
    if (snapshot) {
      for (const key in snapshot) {
        const item = this.childrenType.instantiate(snapshot[key], context, map);
        if (this.childrenType instanceof SafeReferenceType && this.childrenType.options?.acceptsUndefined === false) {
          if (item == null) {
            continue;
          }
        }

        map.set(key, item);
      }
    }

    return map as this["InstanceType"];
  }

  async schemaHash() {
    return `map:${await this.childrenType.schemaHash()}`;
  }
}

export const map = <T extends IAnyType>(childrenType: T): IMapType<T> => {
  ensureRegistered(childrenType);
  return new MapType(childrenType);
};
