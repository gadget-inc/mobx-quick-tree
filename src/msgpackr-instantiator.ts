import { unpack } from "msgpackr";
import type { IClassModelType, IAnyType } from "./types";

export class MsgpackrInstantiatorBuilder<T extends IClassModelType<Record<string, IAnyType>, any, any>> {
  constructor(readonly model: T) {}

  build(): T & { createReadOnlyFromMsgpack: (buffer: Uint8Array) => InstanceType<T> } {
    const originalModel = this.model;
    const enhancedModel = originalModel as any;

    enhancedModel.createReadOnlyFromMsgpack = function (buffer: Uint8Array) {
      const data = unpack(buffer);
      return originalModel.createReadOnly(data);
    };

    return enhancedModel;
  }

  static createFusedInstantiator<T extends IClassModelType<Record<string, IAnyType>, any, any>>(
    model: T,
  ): T & { createReadOnlyFromMsgpack: (buffer: Uint8Array) => InstanceType<T> } {
    const builder = new MsgpackrInstantiatorBuilder(model);
    return builder.build();
  }
}

export function enableMsgpackrIntegration<T extends IClassModelType<Record<string, IAnyType>, any, any>>(
  model: T,
): T & { createReadOnlyFromMsgpack: (buffer: Uint8Array) => InstanceType<T> } {
  return MsgpackrInstantiatorBuilder.createFusedInstantiator(model);
}
