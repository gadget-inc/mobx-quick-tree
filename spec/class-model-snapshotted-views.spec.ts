import { ClassModel, action, snapshottedView, getSnapshot, register, types } from "../src";
import { Apple } from "./fixtures/FruitAisle";
import { create } from "./helpers";

@register
class ViewExample extends ClassModel({ key: types.identifier, name: types.string }) {
  @snapshottedView()
  get slug() {
    return this.name.toLowerCase().replace(/ /g, "-");
  }

  @action
  setName(name: string) {
    this.name = name;
  }
}

@register
class Outer extends ClassModel({ name: types.string, examples: types.array(ViewExample) }) {
  @snapshottedView()
  get upperName() {
    return this.name.toUpperCase();
  }
}

describe("class model snapshotted views", () => {
  // describe.each([
  //   ["read-only", true],
  //   ["observable", false],
  // ])("%s", (_name, readOnly) => {
  describe("whatever", () => {
    const readOnly = false;

    test("instances don't require the snapshot to include the cache", () => {
      const instance = create(ViewExample, { key: "1", name: "Test" }, readOnly);
      expect(instance.slug).toEqual("test");
    });

    test("unchanged instances return the view value in a snapshot of itself when the view is given in the input snapshot", () => {
      const instance = create(ViewExample, { key: "1", name: "Test", slug: "test" } as any, readOnly);
      const snapshot = getSnapshot(instance);
      expect((snapshot as any).slug).toEqual("test");
    });

    test("unchanged instances return the view value in a snapshot of itself when the view is not given in the input snapshot", () => {
      const instance = create(ViewExample, { key: "1", name: "Test" } as any, readOnly);
      const snapshot = getSnapshot(instance);
      expect((snapshot as any).slug).toEqual("test");
    });

    test("snapshots of nested instances returns the view value in the snapshot of itself when not given in the input snapshot", () => {
      const instance = create(Outer, { name: "Test", examples: [{ key: "1", name: "Test" }] } as any, readOnly);

      const snapshot = getSnapshot(instance);
      expect((snapshot as any).upperName).toEqual("TEST");
      expect((snapshot as any).examples[0].slug).toEqual("test");
    });

    test("models with cached views still correctly report .is on totally different models", () => {
      const instance = create(ViewExample, { key: "1", name: "Test" }, readOnly);
      expect(ViewExample.is(instance)).toBe(true);
      expect(Apple.is(instance)).toBe(false);

      const other = create(Apple, { type: "Apple", ripeness: 1 }, readOnly);
      expect(ViewExample.is(other)).toBe(false);
      expect(Apple.is(other)).toBe(true);
    });

    test("instances of models with all optional properties arent .is of other models with all optional properties", () => {
      @register
      class AllOptionalA extends ClassModel({ name: types.optional(types.string, "Jim") }) {}

      @register
      class AllOptionalB extends ClassModel({ title: types.optional(types.string, "Jest") }) {}

      // the empty snapshot matches both types
      expect(AllOptionalA.is({})).toBe(true);
      expect(AllOptionalB.is({})).toBe(true);

      const instanceA = create(AllOptionalA, {}, readOnly);
      expect(AllOptionalA.is(instanceA)).toBe(true);
      expect(AllOptionalB.is(instanceA)).toBe(false);

      const instanceB = create(AllOptionalA, {}, readOnly);
      expect(AllOptionalA.is(instanceB)).toBe(true);
      expect(AllOptionalB.is(instanceB)).toBe(false);
    });

    // blocked on https://github.com/mobxjs/mobx-state-tree/pull/2182
    // test("instances of models with all optional properties arent .is of other models with all optional properties with snapshotted views", () => {
    //   @register
    //   class AllOptionalA extends ClassModel({ name: types.optional(types.string, "Jim") }) {
    //     @snapshottedView()
    //     get slug() {
    //       return this.name.toLowerCase();
    //     }
    //   }

    //   @register
    //   class AllOptionalB extends ClassModel({ title: types.optional(types.string, "Jest") }) {
    //     @snapshottedView()
    //     get whatever() {
    //       return Math.random();
    //     }
    //   }

    //   // the empty snapshot matches both types
    //   expect(AllOptionalA.is({})).toBe(true);
    //   expect(AllOptionalB.is({})).toBe(true);

    //   const instanceA = create(AllOptionalA, {}, readOnly);
    //   expect(AllOptionalA.is(instanceA)).toBe(true);
    //   expect(AllOptionalB.is(instanceA)).toBe(false);

    //   const instanceB = create(AllOptionalA, {}, readOnly);
    //   expect(AllOptionalA.is(instanceB)).toBe(true);
    //   expect(AllOptionalB.is(instanceB)).toBe(false);
    // });
  });

  test("an observable instance saves the view value in a snapshot when changed", () => {
    const instance = ViewExample.create({ key: "1", name: "Test" });
    expect(instance.slug).toEqual("test");
    let snapshot = getSnapshot(instance);
    expect(snapshot).toEqual({ key: "1", name: "Test", slug: "test" });
    instance.setName("New Name");
    snapshot = getSnapshot(instance);
    expect(snapshot).toEqual({ key: "1", name: "New Name", slug: "new-name" });
  });

  test("an observable instance updates the saved view when the observed view value changes", () => {
    const instance = ViewExample.create({ key: "1", name: "Test" });
    instance.setName("New Name");
    expect(instance.slug).toEqual("new-name");
    const snapshot = getSnapshot(instance);
    expect(snapshot).toEqual({ key: "1", name: "New Name", slug: "new-name" });
  });

  test("an observable instance ignores the input snapshot value as the logic may have changed", () => {
    const instance = ViewExample.create({ key: "1", name: "Test", slug: "outdated-cache" } as any);
    expect(instance.slug).toEqual("test");
  });

  test("a readonly instance returns the view value from the snapshot if present", () => {
    const instance = ViewExample.createReadOnly({ key: "1", name: "Test", slug: "test" } as any);
    expect(instance.slug).toEqual("test");
  });

  test("a readonly instance doesn't recompute the view value from the snapshot", () => {
    const instance = ViewExample.createReadOnly({ key: "1", name: "Test", slug: "whatever" } as any);
    expect(instance.slug).toEqual("whatever");
  });

  test("a readonly instance doesn't call the computed function if given a snapshot value", () => {
    const fn = jest.fn();
    @register
    class Spy extends ClassModel({ name: types.string }) {
      @snapshottedView()
      get slug() {
        fn();
        return this.name.toLowerCase().replace(/ /g, "-");
      }
    }

    const instance = Spy.createReadOnly({ name: "Test", slug: "whatever" } as any);
    expect(instance.slug).toEqual("whatever");
    expect(fn).not.toHaveBeenCalled();
  });

  test("an observable instance call the computed function on construction if not given a value", () => {
    const fn = jest.fn();
    @register
    class Spy extends ClassModel({ name: types.string }) {
      @snapshottedView()
      get slug() {
        fn();
        return this.name.toLowerCase().replace(/ /g, "-");
      }
      @action
      setName(name: string) {
        this.name = name;
      }
    }

    const instance = Spy.create({ name: "Test" });
    expect(fn).toHaveBeenCalledTimes(1);

    // not called again upon snapshotting
    getSnapshot(instance);
    expect(fn).toHaveBeenCalledTimes(1);

    instance.setName("New Name");
    expect(fn).toHaveBeenCalled();
  });

  test("snapshotted views can be passed nested within snapshots", () => {
    const instance = Outer.createReadOnly({
      name: "foo",
      upperName: "SNAPSHOT",
      examples: [{ key: "1", name: "Test", slug: "test-foobar" } as any, { key: "2", name: "Test 2", slug: "test-qux" } as any],
    } as any);

    expect(instance.upperName).toEqual("SNAPSHOT");
    expect(instance.examples[0].slug).toEqual("test-foobar");
    expect(instance.examples[1].slug).toEqual("test-qux");
  });

  describe("with a hydrator", () => {
    @register
    class HydrateExample extends ClassModel({ url: types.string }) {
      @snapshottedView<URL>({
        getSnapshot(value, snapshot, node) {
          expect(snapshot).toBeDefined();
          expect(node).toBeDefined();
          return value.toString();
        },
        createReadOnly(value, snapshot, node) {
          expect(snapshot).toBeDefined();
          expect(node).toBeDefined();
          return value ? new URL(value) : undefined;
        },
      })
      get withoutParams() {
        const url = new URL(this.url);
        for (const [key] of url.searchParams.entries()) {
          url.searchParams.delete(key);
        }
        return url;
      }

      @action
      setURL(url: string) {
        this.url = url;
      }
    }

    test("snapshotted views with processors can be accessed on observable instances", () => {
      const instance = HydrateExample.create({ url: "https://gadget.dev/blog/feature?utm=whatever" });
      expect(instance.withoutParams).toEqual(new URL("https://gadget.dev/blog/feature"));
    });

    test("snapshotted views with processors can be accessed on readonly instances when there's no input data", () => {
      const instance = HydrateExample.create({ url: "https://gadget.dev/blog/feature?utm=whatever" });
      expect(instance.withoutParams).toEqual(new URL("https://gadget.dev/blog/feature"));
    });

    test("snapshotted views with processors can be accessed on readonly instances when there is input data", () => {
      const instance = HydrateExample.createReadOnly({
        url: "https://gadget.dev/blog/feature?utm=whatever",
        withoutParams: "https://gadget.dev/blog/feature/extra", // pass a different value so we can be sure it is what is being used
      } as any);
      expect(instance.withoutParams).toEqual(new URL("https://gadget.dev/blog/feature/extra"));
    });
  });

  describe("references", () => {
    @register
    class Referencer extends ClassModel({
      id: types.identifier,
      example: types.reference(ViewExample),
    }) {}

    @register
    class Root extends ClassModel({
      referrers: types.map(Referencer),
      examples: types.map(ViewExample),
    }) {}

    test("references to models with snapshotted views can be instantiated", () => {
      const root = Root.createReadOnly({
        referrers: {
          a: { id: "a", example: "1" },
          b: { id: "b", example: "2" },
        },
        examples: {
          "1": { key: "1", name: "Alice" },
          "2": { key: "2", name: "Bob" },
        },
      });

      expect(getSnapshot(root)).toMatchSnapshot();
    });
  });
});
