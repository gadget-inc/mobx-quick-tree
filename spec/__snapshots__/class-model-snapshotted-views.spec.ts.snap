// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class model snapshotted views references references to models with snapshotted views can be instantiated 1`] = `
{
  "examples": {
    "1": {
      "key": "1",
      "name": "Alice",
      "slug": "alice",
    },
    "2": {
      "key": "2",
      "name": "Bob",
      "slug": "bob",
    },
  },
  "referrers": {
    "a": {
      "example": "1",
      "id": "a",
    },
    "b": {
      "example": "2",
      "id": "b",
    },
  },
}
`;
